<div style="text-align: right"><strong>Capstone #3:</strong> <span style="color:darkred">Supervised Learning</span> </div>

<a id="top"></a>

#### <span style="color:darkred">__Part 1: Data Exploration__ https://github.com/kimrharper/thinkful/blob/master/unit3/unit3-capstone-exploration.ipynb </span><br><br><span style="color:darkred">__Part 2: Models__ https://github.com/kimrharper/thinkful/blob/master/unit3/unit3-capstone-models.ipynb </span>

----

# <span style="color:darkred">Part 2: </span><span style="color:darkblue">L1 Prediction from ELL Writing Samples</span>

__Author:__ Ryan Harper 

----

<a href='#ov'>Overview</a><br>
<a href='#exp'>Experiment</a><br>
<a href='#sec1'>1. Models:</a><br>
><a href='#seca'>A. LR - Ordinary Least Squares</a><br>
<a href='#secb'>B. LR - Logistic Regression</a> <a href='#secb1'> (Lasso)</a> <a href='#secb2'> (Ridge)</a><br>
<a href='#secc'>C. NN - K Nearest Neighbors</a><br>
<a href='#secd'>D. NN - Naive Bayes</a><br>
<a href='#sece'>E. NN - Decision Tree</a><br>
<a href='#secf'>F. Ensemble - Random Forest</a><br>

<a href='#sec2'>2. Model Comparison</a><br>

<a id="ov"></a>

# <span style="color:darkblue">Overview</span>  <a href='#top'>(top)</a>

__Data Source:__
> http://lang-8.com/ [scraped with Beautiful Soup]

![alt text](../data/language/lang8.png "Title")

__Summary:__
> In my previous profession, I have been teaching English to a diverse range of students of all ages, language background, and country origin. During my professional development, I started to observe that different students with different L1s (1st Language) tended to display different patterns of communication that appeared to have some connection to either education in their country of origin or a connection to the linguistic structure of their first language. Different ELL (English Language Learners) needed to focus on different aspects of the English language depending on their background. The purpose of this project is to use a large number of blog posts from a language practicing website and explore whether or not the L1 has any significant impact on the blog writing style of the English learner.<br><br>Part 1: Explore the data to find any noteworthy trends in linguistic structure: <ol><li> vocabulary (word freq, collocations, and cognates) <li>syntax (sentence structure)<li>grammar (i.e. grammar complexity of sentences) <li>errors (types of errors) <li> parts of speech (NLTK Abbreviations: https://pythonprogramming.net/natural-language-toolkit-nltk-part-speech-tagging/)</ol><br>Part 2: Use linguistic trends to determine whether or not a learner's first language can be predicted.

__Variables:__
>__id:__ _User ID_<br>
__time:__ _Time the blog post was scraped (in order of user posted time)_ <br>
__title:__ _Title of the blog post_<br>
__content:__ _The blog post_<br>
__language:__ _User's self-reported first language_

<a id="exp"></a>

# <span style="color:darkblue">Experiment</span> <a href='#top'>(top)</a>

__Hypothesis:__ 
> L1 (first language) experience and academic environment influences ELLs' (English Language Learners') writing style. The L1 of ELLs can be predicted by looking at English blog posts and identifying patterns unique to their L1.

__Observations:__
><li> --<li>--<li>--

__Method:__
> Using multiple different models. The aim of this project is to explore how different models can handle the data (target and features) and to see what information can be gained from using multiple different models. Ultimately, the goal is to determine which models are appropriate for a binary (discrete) target with features that are both qualitative (discrete) and quantitative (ranked/continuous).

<a id="sec1"></a>

# <span style="color:darkblue">1. Models:</span>  <a href='#top'>(top)</a>

In [1]:
# iPython/Jupyter Notebook
import time
from pprint import pprint
import warnings
from IPython.display import Image

import time

# Data processing
import pandas as pd
import plotly as plo
import seaborn as sns
from scipy import stats
from collections import Counter
import numpy as np
import itertools

# NLP
from nltk.corpus import stopwords as sw
from nltk.util import ngrams
from nltk.corpus import brown
import nltk
import re
from nltk.tokenize import RegexpTokenizer
import difflib

# Stats
from sklearn.metrics import classification_report, roc_curve,roc_auc_score,accuracy_score
from sklearn import metrics

# Preparing Models
from sklearn.model_selection import train_test_split

# Decomposition
from sklearn.decomposition import TruncatedSVD
from sklearn.random_projection import sparse_random_matrix

# Models
from sklearn import linear_model
from sklearn.neighbors import KNeighborsClassifier
from sklearn import tree
from sklearn.naive_bayes import BernoulliNB,MultinomialNB,GaussianNB

# Ensemble
from sklearn import ensemble
from sklearn.model_selection import cross_val_score

#Visualization
from IPython.display import Image
import pydotplus
import graphviz

# import altair as alt

__Import Features + Target__

In [39]:
features = pd.read_csv('blogfeatures.csv').sample(frac=1.0)
del features['Unnamed: 0']
del features['id']
lang = list(features.language.unique())

__Declare X,y Variables__

In [40]:
y = features['language'].values.reshape(-1, 1).ravel()
X = features[features.columns[~features.columns.str.contains('language')]]
X.head()

print(np.shape(y))
print(np.shape(X))

(14148,)
(14148, 14665)


__SVD Truncate: To find most important features__

In [41]:
svd = TruncatedSVD(n_components=20, n_iter=7, random_state=42)
svd.fit(X)  

TruncatedSVD(algorithm='randomized', n_components=20, n_iter=7,
       random_state=42, tol=0.0)

In [63]:
best_features = [X.columns[i] for i in svd.components_[0].argsort()[::-1]][0:50]

__Reduce Features (Unused): 14,665 to n_components __

__Split Data to Train/Test__

In [64]:
X_train, X_test, y_train, y_test = train_test_split(X[best_features], y, test_size=0.20, random_state=32)  

<a id="seca"></a>

__Create Function for Comparing Models__

In [65]:
cols = ['name','time','total','precision','recall','f1']

model_set = pd.DataFrame(columns=cols)
models_stored = []
pattern = "%.2f"

In [66]:
def run_model(model,name):
    global model_set
    m = model
    m.fit(X_train, y_train)
    start = time.time()

    total_score = m.score(X_test,y_test)
    pscore = [pattern % i for i in list(metrics.precision_score(y_test, m.predict(X_test),labels=lang,average=None))]
    rscore = [pattern % i for i in list(metrics.recall_score(y_test, m.predict(X_test),labels=lang,average=None))]
    fscore = [pattern % i for i in list(metrics.f1_score(y_test, m.predict(X_test),labels=lang,average=None))]
    end = time.time()
    t= pattern % (end - start)

    r = dict(zip(cols,[name,t,total_score,pscore,rscore,fscore]))
    print('Total Score is: {}\n'.format(total_score))
    print(classification_report(y_test, m.predict(X_test)))
    
    model_set = model_set.append(r,ignore_index=True)
    return r,m

### <span style="color:gray">A. LR - Ordinary Least Squares _(not used)_</span>  <a href='#top'>(top)</a>

> Target is discrete so this model may not be appropriate <br>Many features are binary so model may not be appropriate

<a id="secb"></a>

### <span style="color:darkred">B. LR - Logistic Regression</span>  <a href='#top'>(top)</a>

> Target is binary so logistic regression will operate on probabilities

In [67]:
%%time
lreg_data,lreg = run_model(linear_model.LogisticRegression(),'Logistic Regression')

Total Score is: 0.7219081272084805

                     precision    recall  f1-score   support

            English       0.00      0.00      0.00        15
           Japanese       0.73      0.95      0.83      1906
             Korean       0.00      0.00      0.00       241
Traditional Chinese       0.64      0.36      0.46       668

        avg / total       0.65      0.72      0.67      2830

CPU times: user 972 ms, sys: 31.8 ms, total: 1 s
Wall time: 1.06 s


In [68]:
pd.crosstab(y_test,lreg.predict(X_test))

col_0                English  Japanese  Korean  Traditional Chinese
row_0                                                              
English                    0        10       0                    5
Japanese                   0      1805       1                  100
Korean                     0       214       0                   27
Traditional Chinese        1       429       0                  238

In [69]:
# print(np.std(X)*lreg.coef_)

<a id="secb1"></a>

### <span style="color:gray">C. Lasso _(not used)_</span>  <a href='#top'>(top)</a>

<a id="secb2"></a>

### <span style="color:gray">D. Ridge _(not used)_</span>  <a href='#top'>(top)</a>

_Lasso and Ridge are not good predictors so should I just be using them for parameter manipulation?_

<a id="sece"></a>

### <span style="color:darkred">E. K Nearest Neighbors</span>  <a href='#top'>(top)</a>

> Can handle discrete values for target <br>Quantitative values are limited (not continuous) and might be problematic for nearest neighbors

In [70]:
%%time
neighbors_data,neighbors = run_model(KNeighborsClassifier(n_neighbors=5),'K Nearest Neighbor')

/Users/ryan/anaconda3/lib/python3.6/site-packages/sklearn/metrics/classification.py:1135: UndefinedMetricWarning:

Precision is ill-defined and being set to 0.0 in labels with no predicted samples.

/Users/ryan/anaconda3/lib/python3.6/site-packages/sklearn/metrics/classification.py:1135: UndefinedMetricWarning:

F-score is ill-defined and being set to 0.0 in labels with no predicted samples.



Total Score is: 0.684452296819788

                     precision    recall  f1-score   support

            English       0.00      0.00      0.00        15
           Japanese       0.72      0.92      0.81      1906
             Korean       0.13      0.02      0.04       241
Traditional Chinese       0.52      0.28      0.36       668

        avg / total       0.62      0.68      0.63      2830

CPU times: user 1.77 s, sys: 17.7 ms, total: 1.79 s
Wall time: 1.83 s


/Users/ryan/anaconda3/lib/python3.6/site-packages/sklearn/metrics/classification.py:1135: UndefinedMetricWarning:

Precision and F-score are ill-defined and being set to 0.0 in labels with no predicted samples.



<a id="secf"></a>

### <span style="color:darkred">F. Naive Bayes - Bernoulli</span>  <a href='#top'>(top)</a>

> Should be best for boolean classification but not for multiple discrete values (multinomial should be better)

In [71]:
%%time
bnb_data,bnb = run_model(BernoulliNB(),'Naive Bayes - Bernoulli')

Total Score is: 0.6123674911660777

                     precision    recall  f1-score   support

            English       0.00      0.00      0.00        15
           Japanese       0.71      0.78      0.75      1906
             Korean       0.00      0.00      0.00       241
Traditional Chinese       0.33      0.36      0.34       668

        avg / total       0.56      0.61      0.58      2830

CPU times: user 110 ms, sys: 17.8 ms, total: 128 ms
Wall time: 134 ms


/Users/ryan/anaconda3/lib/python3.6/site-packages/sklearn/metrics/classification.py:1135: UndefinedMetricWarning:

Precision is ill-defined and being set to 0.0 in labels with no predicted samples.

/Users/ryan/anaconda3/lib/python3.6/site-packages/sklearn/metrics/classification.py:1135: UndefinedMetricWarning:

F-score is ill-defined and being set to 0.0 in labels with no predicted samples.

/Users/ryan/anaconda3/lib/python3.6/site-packages/sklearn/metrics/classification.py:1135: UndefinedMetricWarning:

Precision and F-score are ill-defined and being set to 0.0 in labels with no predicted samples.



<a id="secg"></a>

### <span style="color:darkred">G. Decision Tree</span>  <a href='#top'>(top)</a>

> Visualizes most important features by hierarchy <br>Longer processing time

In [72]:
%%time
dt_data,dt = run_model(tree.DecisionTreeClassifier(criterion='entropy',max_depth=7),'Decision Tree')

Total Score is: 0.7060070671378091

                     precision    recall  f1-score   support

            English       0.50      0.07      0.12        15
           Japanese       0.72      0.94      0.82      1906
             Korean       0.00      0.00      0.00       241
Traditional Chinese       0.59      0.30      0.40       668

        avg / total       0.63      0.71      0.65      2830

CPU times: user 141 ms, sys: 9.9 ms, total: 151 ms
Wall time: 159 ms


In [74]:
# Render tree.
dot_data = tree.export_graphviz(
    dt, out_file=None,
    feature_names=best_features,
    class_names=lang,
    filled=True
)

graph = pydotplus.graph_from_dot_data(dot_data)
Image(graph.create_png())

graph.write_png('decision_tree.png')

True

In [75]:
dimportance = list(zip(best_features[0:20],dt.feature_importances_))
dimportance = dict(dimportance)
a1_sorted_keys = sorted(dimportance, key=dimportance.get, reverse=True)
p = []
for r in a1_sorted_keys:
    if dimportance[r] != 0:
        p.append(r)
#         print(r, dimportance[r])
        
print(p)

['sc', 'pos2_DT-JJ', 'wc', 'pos2_PRP-VBD', 'pos2_NN-NN', 'pos2_PRP-VBP', 'pos3_IN-DT-NN', 'pos3_DT-NN-IN', 'pos2_IN-NN', 'pos2_IN-PRP', 'pos2_DT-NN', 'pos2_NN-CC', 'pos2_NNS-IN', 'pos2_JJ-NN', 'pos2_TO-VB', 'pos2_NN-IN', 'pos2_IN-DT']


_Good visualization of important features and presentation of entropy weighting_

<a id="sech"></a>

### <span style="color:darkred">H. Random Forest</span>  <a href='#top'>(top)</a>

> Runs decision tree multiple times for best output <br>Longest processing time

In [76]:
%%time
rf_data,rf = run_model(ensemble.RandomForestClassifier(n_estimators=20),'Random Forest')

Total Score is: 0.7173144876325088

                     precision    recall  f1-score   support

            English       0.00      0.00      0.00        15
           Japanese       0.73      0.95      0.82      1906
             Korean       0.00      0.00      0.00       241
Traditional Chinese       0.64      0.33      0.44       668

        avg / total       0.64      0.72      0.66      2830

CPU times: user 528 ms, sys: 24.5 ms, total: 553 ms
Wall time: 562 ms


/Users/ryan/anaconda3/lib/python3.6/site-packages/sklearn/metrics/classification.py:1135: UndefinedMetricWarning:

Precision is ill-defined and being set to 0.0 in labels with no predicted samples.

/Users/ryan/anaconda3/lib/python3.6/site-packages/sklearn/metrics/classification.py:1135: UndefinedMetricWarning:

F-score is ill-defined and being set to 0.0 in labels with no predicted samples.

/Users/ryan/anaconda3/lib/python3.6/site-packages/sklearn/metrics/classification.py:1135: UndefinedMetricWarning:

Precision and F-score are ill-defined and being set to 0.0 in labels with no predicted samples.



In [77]:
cvs = cross_val_score(rf, X_test, y_test, cv=5)
print(cvs)
scoreH = cvs.sum()/len(cvs)

[0.71126761 0.73144876 0.69964664 0.70619469 0.70088496]


In [78]:
rf.score(X_train,y_train)
print(scoreH)

0.7098885316023853


<a id="sec2"></a>

# <span style="color:darkblue">2. Model Comparison</span>  <a href='#top'>(top)</a>

In [80]:
model_set.columns = ['name','time','total','prec: | JA | CH | KO | EN |','rec: | JA | CH | KO | EN |','f1: | JA | CH | KO | EN |']
model_set

name  time     total prec: | JA | CH | KO | EN |  \
0      Logistic Regression  0.05  0.721908    [0.73, 0.64, 0.00, 0.00]   
1       K Nearest Neighbor  1.43  0.684452    [0.72, 0.52, 0.13, 0.00]   
2  Naive Bayes - Bernoulli  0.05  0.612367    [0.71, 0.33, 0.00, 0.00]   
3            Decision Tree  0.04  0.706007    [0.72, 0.59, 0.00, 0.50]   
4            Random Forest  0.10  0.717314    [0.73, 0.64, 0.00, 0.00]   

  rec: | JA | CH | KO | EN | f1: | JA | CH | KO | EN |  
0   [0.95, 0.36, 0.00, 0.00]  [0.83, 0.46, 0.00, 0.00]  
1   [0.92, 0.28, 0.02, 0.00]  [0.81, 0.36, 0.04, 0.00]  
2   [0.78, 0.36, 0.00, 0.00]  [0.75, 0.34, 0.00, 0.00]  
3   [0.94, 0.30, 0.00, 0.07]  [0.82, 0.40, 0.00, 0.12]  
4   [0.95, 0.33, 0.00, 0.00]  [0.82, 0.44, 0.00, 0.00]

-----

Regularization